fud2/fud-core/src/exec/driver.rs

```diff
@@ -25,104 +24,176 @@ pub struct Driver {
+            {
+                return Some(plan);
+            }
+        }
+        None
     }
 
     /// Generate a filename with an extension appropriate for the given State.
-    fn gen_name(&self, stem: &str, state: StateRef) -> Utf8PathBuf {
-        let state = &self.states[state];
-        if state.is_pseudo() {
-            Utf8PathBuf::from(format!("_pseudo_{}", state.name))
+    /// This function assumes all input states are unique and all output states are unique.
+    /// Additionally, it assume all non-input/output states are unique.
+    fn gen_name(
+        &self,
+        state_ref: StateRef,
+        used: bool,
+        req: &Request,
+        used_as_input: &mut HashSet<StateRef>,
+    ) -> IO {
+        let state = &self.states[state_ref];
+        let extension = if !state.extensions.is_empty() {
+            &state.extensions[0]
         } else {
-            // TODO avoid collisions in case we reuse extensions...
-            Utf8PathBuf::from(stem).with_extension(&state.extensions[0])
+            ""
+        };
+
+        // make sure we have correct input/output filenames and mark if we read from stdio
+        if req.start_state.contains(&state_ref)
+            && !used_as_input.contains(&state_ref)
+        {
+            used_as_input.insert(state_ref);
+            let idx = req.start_state.partition_point(|&r| r == state_ref) - 1;
+            if let Some(start_file) = req.start_file.get(idx) {
+                return IO::File(utils::relative_path(
+                    start_file,
+                    &req.workdir,
+                ));
+            } else {
+                return IO::StdIO(
+                    idx,
+                    utils::relative_path(
+                        &Utf8PathBuf::from(format!(
+                            "_from_stdin_{}",
+                            state.name,
+                        ))
+                        .with_extension(extension),
+                        &req.workdir,
+                    ),
+                );
+            }
+        }
+
+        if req.end_state.contains(&state_ref) {
+            let idx = req.end_state.partition_point(|&r| r == state_ref) - 1;
+            if let Some(end_file) = req.end_file.get(idx) {
+                return IO::File(utils::relative_path(end_file, &req.workdir));
+            } else {
+                return IO::StdIO(
+                    idx,
+                    utils::relative_path(
+                        &Utf8PathBuf::from(format!(
+                            "_to_stdout_{}",
+                            state.name
+                        ))
+                        .with_extension(extension),
+                        &req.workdir,
+                    ),
+                );
+            }
         }
+
+        // okay, wild west filename time (make them unique and hopefully helpful)
+        let prefix = if !used { "_unused_" } else { "" };
```

Comment:
It is possible that an output of an op is not used. For example, an op could generate two outputs, but only one is used. It would then be nice for a different op to still be able to generate a file of that unused output's state, otherwise a perfectly feasible plan maybe not be able to run due to a conflict of names.